### PR TITLE
Implement Setter API

### LINKED ISSUE #329: Add "builder style" setter APIs that permit setting durability
The `salsa::input` structs generate a setter for each field. We should make an alternate version of the setter that takes a durability. The obvious thing is something like `set_foo_with_durability`, but now that I think about it, maybe a builder API would be more appropriate, in order to be more future proof.

I'm inclined to modify the setters to be builder style. You would write `input.set_foo(&mut db)` and that would return a `salsa::Setter` struct that captures the input, a reference to the field (a `FunctionIngredient`), and the database. The setter would have a `to_value(self, value)` that stores the new value, so that the "normal" call looks like

```rust
input.set_foo(&mut db).to_value(value)
```

but then we can add additional builder style methods, e.g.

```rust
input.set_foo(&mut db).with_durability(Durability::HIGH).to_value(value)
```

This makes those method highly discoverable and easily documented. It also means we can add new "final actions", e.g. maybe for "taking", deleting, or something else in the future.

### PR DESCRIPTION
**Pull request by crlf0710**
Fixes #329

### CODE REVIEW

File: components/salsa-2022/src/setter.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            ingredient,
+            durability: Durability::LOW,
+        }
+    }
+
+    pub fn with_durability(self, durability: Durability) -> Self {
+        Setter { durability, ..self }
+    }
+
+    pub fn to_value(self, value: F) -> F {
```
Comment (nikomatsakis) [inline components/salsa-2022/src/setter.rs:34]: Nit: maybe we should just call it `to`, just to keep the simplest case short? 

e.g.

```rust
func.set_name(db).to(value)
```

File: components/salsa-2022/src/setter.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    ) -> Self {
+        Setter {
+            runtime,
+            key,
+            ingredient,
+            durability: Durability::LOW,
+        }
+    }
+
+    pub fn with_durability(self, durability: Durability) -> Self {
```
Comment (nikomatsakis) [inline components/salsa-2022/src/setter.rs:30]: Nit: we don't have any tests of durability! Maybe you want to add one? That said, I'm ok to land this PR and port the older tests over as follow-up.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks great! I left one :bike:-shed question (`to` instead of `to_value`), but the code itself looks ready to land. Do you think you'd be up to port over some of the durability tests from old salsa?

File: components/salsa-2022/src/setter.rs (hunk 1 above)
Comment (crlf0710) [reply components/salsa-2022/src/setter.rs]: Done~

**Comment by crlf0710** [comment]
Rebased and renamed `to_value` to `to` following review.

File: components/salsa-2022/src/setter.rs (hunk 2 above)
Comment (crlf0710) [reply components/salsa-2022/src/setter.rs:30]: That's #339 so I'd like to leave it to a follow-up PR. I'll see if i can get it right.

**Comment by nikomatsakis** [comment]
Neat.

bors r+

**Review by nikomatsakis — APPROVED** [review]

**Comment by crlf0710** [comment]
Oops, there's one left-over in the calc example. Fixed now.

**Pull request merged**